### PR TITLE
feat: refactor core module into multiple packages

### PR DESCRIPTION
**Pull request by thisthat**
Signed-off-by: Giovanni Liva <giovanni.liva@dynatrace.com>

## This PR

- Refactor the service_[connect|metrics].go into multiple files 
- Adds more unit tests when easy and feasible 

### Related Issues

Preparation for #501

### CODE REVIEW

**Comment by toddbaert** [comment]
I like this refactor. In general it's my preference to see refactors done in separate PRs. It's helps to be able to approach a refactor PR knowing there's no functional changes. I'd like to see it separated if possible, but it's not necessary.

**Comment by thisthat** [comment]
> I like this refactor. In general it's my preference to see refactors done in separate PRs. It's helps to be able to approach a refactor PR knowing there's no functional changes. I'd like to see it separated if possible, but it's not necessary.

That's a fair argument. I will remove the impressions and just tackle the refactoring with this PR 👍

File: core/pkg/otel/metrics.go [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+func (r MetricsRecorder) InFlightRequestEnd(ctx context.Context, attrs []attribute.KeyValue) {
+	r.httpRequestsInflight.Add(ctx, -1, attrs...)
+}
+
+func (r MetricsRecorder) Impressions(ctx context.Context, key, variant string) {
+	r.impressions.Add(ctx, 1, []attribute.KeyValue{
+		semconv.FeatureFlagKey(key),
+		semconv.FeatureFlagVariant(variant),
+		semconv.FeatureFlagProviderName("flagd"),
+	}...)
```
Comment (toddbaert) [inline core/pkg/otel/metrics.go:54]: Seems right!

File: core/pkg/otel/metrics.go [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+}
+
+func NewOTelRecorder(exporter metric.Reader, serviceName string) *MetricsRecorder {
+	const requestDurationName = "http_request_duration_seconds"
+	const responseSizeName = "http_response_size_bytes"
+
+	// create a metric provider with custom bucket size for histograms
+	provider := metric.NewMeterProvider(
+		metric.WithReader(exporter),
+		metric.WithView(getDurationView(requestDurationName, serviceName, prometheus.DefBuckets)),
```
Comment (toddbaert) [inline core/pkg/otel/metrics.go:79]: Do we have to use prometheus APIs instead of OTel for these metrics?

Comment (thisthat) [reply core/pkg/otel/metrics.go:79]: This is for just keeping the same bucket size as it was before moving to OTel as SDK to report metrics (see https://github.com/open-feature/flagd/pull/419#issuecomment-1440806008)

File: core/pkg/otel/metrics.go [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+		metric.WithView(getDurationView(responseSizeName, serviceName, prometheus.ExponentialBuckets(100, 10, 8))),
```
[the remark this answers is not in the document]
Comment (thisthat) [reply core/pkg/otel/metrics.go:80]: I don't think a constant make sense here, but definitely a comment helps the reader to understand why we change the bucket size :)

File: core/pkg/otel/metrics.go (hunk 1 above)
Comment (thisthat) [reply core/pkg/otel/metrics.go:54]: As discussed before, I'll remove this and tackle the introduction of the new metric in a separate PR - this way the changelog is also nicer :)

File: core/pkg/otel/metrics.go (hunk 2 above)
Comment (thisthat) [reply core/pkg/otel/metrics.go:79]: Maybe to expand more on this subject: using OTel allows us to be vendor agnostic and provide metrics in different formats. The "output format" currently is wired inside the [`from_config` file](https://github.com/open-feature/flagd/pull/530/files#diff-60ffee6ef9753f574677d132f3e595feee457c5e52c1b1b88063c616a3c6cdf7R57)

File: core/pkg/otel/metrics_test.go [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+				semconv.HTTPMethodKey.String(""),
+				semconv.HTTPStatusCodeKey.String(""),
+			},
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			rec := MetricsRecorder{}
+			res := rec.HTTPAttributes(tt.req.Service, tt.req.ID, tt.req.Method, tt.req.Code)
+			if len(res) != 4 {
```
Comment (bacherfl) [inline core/pkg/otel/metrics_test.go:78]:
```suggestion
require.Equal(t, tt.want, res)
```

**Comment by thisthat** [comment]
I need some help with the integration-tests @toddbaert 
Locally, they also don't work. The strange thing is that I can reach the binary under test via curl `curl -X POST "localhost:8013/schema.v1.Service/ResolveString" -d '{"flagKey":"string-flag","context":{}}' -H "Content-Type: application/json"`

**Comment by thisthat** [comment]
> I need some help with the integration-tests @toddbaert Locally, they also don't work. The strange thing is that I can reach the binary under test via curl `curl -X POST "localhost:8013/schema.v1.Service/ResolveString" -d '{"flagKey":"string-flag","context":{}}' -H "Content-Type: application/json"`

For the record, connect-go requires an implementation of `Flush()` to work properly

**Review by bacherfl — APPROVED** [review]
lgtm

**Review by toddbaert — APPROVED** [review]

**Pull request merged**